sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveWithCTE.scala

```diff
@@ -41,16 +47,113 @@ object ResolveWithCTE extends Rule[LogicalPlan] {
+                  cteDef
+                } else {
+                  val loop = UnionLoop(
+                    cteDef.id,
+                    anchor,
+                    rewriteRecursiveCTERefs(recursion, anchor, cteDef.id, None))
+                  cteDef.copy(child = alias.copy(child = loop))
+                }
+
+              // The case of CTE name followed by a parenthesized list of column name(s), eg.
+              // WITH RECURSIVE t(n).
+              case alias @ SubqueryAlias(_,
+              columnAlias @ UnresolvedSubqueryColumnAliases(
+              colNames,
+              Union(Seq(anchor, recursion), false, false)
+              )) =>
+                if (!anchor.resolved) {
+                  cteDef
+                } else {
+                  val loop = UnionLoop(
+                    cteDef.id,
+                    anchor,
+                    rewriteRecursiveCTERefs(recursion, anchor, cteDef.id, Some(colNames)))
+                  cteDef.copy(child = alias.copy(child = columnAlias.copy(child = loop)))
+                }
+
+              // If the recursion is described with an UNION (deduplicating) clause then the
+              // recursive term should not return those rows that have been calculated previously,
+              // and we exclude those rows from the current iteration result.
+              case alias @ SubqueryAlias(_,
+              Distinct(Union(Seq(anchor, recursion), false, false))) =>
+                if (!anchor.resolved) {
+                  cteDef
+                } else {
+                  val loop = UnionLoop(
+                    cteDef.id,
+                    Distinct(anchor),
+                    Except(
+                      rewriteRecursiveCTERefs(recursion, anchor, cteDef.id, None),
+                      UnionLoopRef(cteDef.id, anchor.output, true),
+                      isAll = false
+                    )
+                  )
+                  cteDef.copy(child = alias.copy(child = loop))
+                }
+
+              // The case of CTE name followed by a parenthesized list of column name(s).
+              case alias @ SubqueryAlias(_,
+              columnAlias@UnresolvedSubqueryColumnAliases(
+              colNames,
+              Distinct(Union(Seq(anchor, recursion), false, false))
+              )) =>
+                if (!anchor.resolved) {
+                  cteDef
+                } else {
+                  val loop = UnionLoop(
+                    cteDef.id,
+                    Distinct(anchor),
+                    Except(
+                      rewriteRecursiveCTERefs(recursion, anchor, cteDef.id, Some(colNames)),
+                      UnionLoopRef(cteDef.id, anchor.output, true),
+                      isAll = false
+                    )
+                  )
+                  cteDef.copy(child = alias.copy(child = columnAlias.copy(child = loop)))
+                }
+
+              case other if !other.exists(_.isInstanceOf[UnionLoop]) =>
```

Comment:
```suggestion
              case other =>
```
We always rewrite Union -> UnionLoop and recursive ref -> UnionLoopRef together. Once the rewriting is done, we will hit the non-recursive branch.